### ISSUE TITLE
postfix projections not working for (non-field) functions

### ISSUE THREAD

**Issue by m0davis** [agda/agda#2587]
Postfix projections are not working when the projected function is not a field of some record module. Agda rejects the following:

```agda
record R : Set where
  foo : Set₁
  foo = Set

test-foo : R → Set₁
test-foo r = r .R.foo
```
The error is:
```
R should be a function type, but it isn't
when checking that .R.foo is a valid argument to a function of type
R
```
Fwiw, such projections do work when the function is a non-field of the given record module (but is a field somewhere):
```agda
record S : Set₂ where
  field bar : Set₁

record T : Set₂ where
  field s : S
  open S s public

test-bar : T → Set₁
test-bar t = t .T.bar
```

**Comment by andreasabel** [comment]
In contrast to `bar`, for the scope checker `foo` is not a projection, thus, the postfix machinery does not get started.  

We could treat record methods such as `foo` special and allow postfix projection for them.  Would make Agda more like object-style languages.  However, I think we should not allow postfix notation for arbitrary functions.

I suppose you expect post-fix notation to work for all record declarations, do you?
```agda
module _ (A : Set) where

  record R : Set where
    foo : Set
    foo = A

    data D : Set where
      c : D

test-foo : ∀ A → R A → Set
test-foo A r = r .R.foo

test-D : ∀ A (r : R A) → (r .R.D)
test-D A r = r .R.c
```
If we also want faithful printing for this situation, constructors and functions in general would have to store whether they come from a postfix application or a prefix one.

**Comment by m0davis** [comment]
@andreasabel, yes, your supposition about my expectation is correct. I (perhaps naively) think of the postfix stuff as simply an "alternate spelling" of prefix stuff.

P.s. It's only by accident that I noticed your updated comment. Github does not notify via email when a conversation is updated via an edit, so it's best to only edit for clarity and put new content in a separate comment.

**Comment by JasonGross** [comment]
I'd be interested in having this too.  What's the reason it was closed as wontfix?

**Comment by andreasabel** [comment]
> I'd be interested in having this too. What's the reason it was closed as wontfix?

Not sure.  We can keep this issue open, but would likely need a volunteer to implement this.
